### PR TITLE
Add failing test cases for generative interfaces and abstract classes #211

### PR DESCRIPTION
**Pull request by panesofglass**
Add test cases for #211:
- [ ] generated interfaces
- [ ] generated abstract classes with abstract methods
- [x] generated abstract classes with virtual methods

### CODE REVIEW

**Comment by sergey-tihon** [comment]
There is a small chance that we have to manually set attributes like `hidebysig`,`newslot`, `auto`, `ansi`

File: tests/GenerativeAbstractClassesTests.fs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        |> List.map (fun (name, parameters, retType, isVirtual) ->
+            let ps =
+                parameters
+                |> List.map (fun (name, ty) ->
+                    ProvidedParameter(name, ty))
+            if isVirtual then
+                let m = ProvidedMethod(name, ps, retType, invokeCode = fun args ->
+                    <@ raise (NotImplementedException(name + " is not implemented")) @>.Raw
+                    )
+                m.SetMethodAttrs (MethodAttributes.Public ||| MethodAttributes.HideBySig ||| MethodAttributes.NewSlot ||| MethodAttributes.Abstract ||| MethodAttributes.Virtual)
```
Comment (panesofglass) [inline tests/GenerativeAbstractClassesTests.fs]: Not sure these are correct. However, the build still fails on the interfaces, so I'm not sure.

**Comment by dsyme** [comment]
@panesofglass These are good tests, thanks.  I think I can look after #212 from here and get these tests passing

**Comment by panesofglass** [comment]
Thank you, @dsyme!

**Comment by dsyme** [comment]
@panesofglass I know I've promised you to get to this, but I won't be able to get to it until the perf problem in #220 is fixed, which may take me a while

**Comment by 7sharp9** [comment]
Whats the current issue with generated interfaces?

**Comment by panesofglass** [comment]
@7sharp9 it seems the metadata is not currently generated correctly.

**Comment by panesofglass** [comment]
I just updated and tried again. This _seems_ to be reporting better than what I recall:

```
[xUnit.net 00:00:04.4116552]     FSharp.TypeProviders.SDK.Tests.GenerativeAbstractClassesTests.Abstract classes are generated correctly [FAIL]
[xUnit.net 00:00:04.4329294]     FSharp.TypeProviders.SDK.Tests.GenerativeInterfacesTests.Interfaces are generated correctly [FAIL]
Failed   FSharp.TypeProviders.SDK.Tests.GenerativeAbstractClassesTests.Abstract classes are generated correctly
Error Message:
 System.Exception : The provided method is marked as an abstract method; therefore, it should not define an implementation.
Stack Trace:
   at <StartupCode$FSharp-TypeProviders-SDK-Tests>.$ProvidedTypes.Compile@14379-14.Invoke(ILTypeBuilder tb, FSharpOption`1 ptdT) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14481
   at <StartupCode$FSharp-TypeProviders-SDK-Tests>.$ProvidedTypes.typeMembers@14214(AssemblyCompiler this, FSharpFunc`2 f, ProvidedTypeDefinition ptd) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14216
   at <StartupCode$FSharp-TypeProviders-SDK-Tests>.$ProvidedTypes.typeMembers@14214(AssemblyCompiler this, FSharpFunc`2 f, ProvidedTypeDefinition ptd) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14218
   at ProviderImplementation.ProvidedTypes.AssemblyCompiler.iterateTypes[a](FSharpFunc`2 f, IEnumerable`1 providedTypeDefinitions) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14228
   at <StartupCode$FSharp-TypeProviders-SDK-Tests>.$ProvidedTypes.Compile@14379-13.Invoke(IEnumerable`1 providedTypeDefinitions) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14379
   at ProviderImplementation.ProvidedTypes.AssemblyCompiler.Compile(Boolean isHostedExecution) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14263
   at ProviderImplementation.ProvidedTypes.TypeProviderForNamespaces.Microsoft-FSharp-Core-CompilerServices-ITypeProvider-GetGeneratedAssemblyContents(Assembly assembly) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14794
   at FSharp.TypeProviders.SDK.Tests.GenerativeAbstractClassesTests.testProvidedAssembly(FSharpFunc`2 test) in C:\Code\FSharp.TypeProviders.SDK\tests\GenerativeAbstractClassesTests.fs:line 81
Failed   FSharp.TypeProviders.SDK.Tests.GenerativeInterfacesTests.Interfaces are generated correctly
Error Message:
 System.Exception : The provided type definition is an interface; therefore, it should not define an implementation for its members.
Stack Trace:
   at <StartupCode$FSharp-TypeProviders-SDK-Tests>.$ProvidedTypes.Compile@14379-14.Invoke(ILTypeBuilder tb, FSharpOption`1 ptdT) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14479
   at <StartupCode$FSharp-TypeProviders-SDK-Tests>.$ProvidedTypes.typeMembers@14214(AssemblyCompiler this, FSharpFunc`2 f, ProvidedTypeDefinition ptd) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14216
   at <StartupCode$FSharp-TypeProviders-SDK-Tests>.$ProvidedTypes.typeMembers@14214(AssemblyCompiler this, FSharpFunc`2 f, ProvidedTypeDefinition ptd) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14218
   at ProviderImplementation.ProvidedTypes.AssemblyCompiler.iterateTypes[a](FSharpFunc`2 f, IEnumerable`1 providedTypeDefinitions) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14228
   at <StartupCode$FSharp-TypeProviders-SDK-Tests>.$ProvidedTypes.Compile@14379-13.Invoke(IEnumerable`1 providedTypeDefinitions) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14379
   at ProviderImplementation.ProvidedTypes.AssemblyCompiler.Compile(Boolean isHostedExecution) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14263
   at ProviderImplementation.ProvidedTypes.TypeProviderForNamespaces.Microsoft-FSharp-Core-CompilerServices-ITypeProvider-GetGeneratedAssemblyContents(Assembly assembly) in C:\Code\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14794
   at FSharp.TypeProviders.SDK.Tests.GenerativeInterfacesTests.testProvidedAssembly(FSharpFunc`2 test) in C:\Code\FSharp.TypeProviders.SDK\tests\GenerativeInterfacesTests.fs:line 71

Total tests: 39. Passed: 37. Failed: 2. Skipped: 0.
Test Run Failed.
Test execution time: 9.4491 Seconds
Running build failed.
Error:
System.Exception: Test failed on "test" "tests/FSharp.TypeProviders.SDK.Tests.fsproj" --configuration Release --framework net461
   at Fake.DotNetCli.Test@298-3.Invoke(String message) in D:\code\fake\src\app\FakeLib\DotNetCLIHelper.fs:line 298
   at Fake.DotNetCli.Test(FSharpFunc`2 setTestParams) in D:\code\fake\src\app\FakeLib\DotNetCLIHelper.fs:line 298
   at FSI_0005.Build.clo@116-19.Invoke(Unit _arg4) in C:\Code\FSharp.TypeProviders.SDK\build.fsx:line 135
   at Fake.TargetHelper.runSingleTarget(TargetTemplate`1 target) in D:\code\fake\src\app\FakeLib\TargetHelper.fs:line 626
```

**Comment by panesofglass** [comment]
After updating to leverage optional `invokeCode` and fixing a minor issue with correctly inferring requirement for virtual members, the failing tests produce:

```
[xUnit.net 00:00:12.8257338]     FSharp.TypeProviders.SDK.Tests.GenerativeInterfacesTests.Interfaces are generated correctly [FAIL]
[xUnit.net 00:00:12.8981477]     FSharp.TypeProviders.SDK.Tests.GenerativeAbstractClassesTests.Abstract classes are generated correctly [FAIL]
Failed   FSharp.TypeProviders.SDK.Tests.GenerativeInterfacesTests.Interfaces are generated correctly
Error Message:
 System.TypeLoadException : Failure has occurred while loading a type.
Stack Trace:
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at System.Reflection.Assembly.get_ExportedTypes()
   at FSharp.TypeProviders.SDK.Tests.GenerativeInterfacesTests.testProvidedAssembly(FSharpFunc`2 test) in C:\Code\FSharp.TypeProviders.SDK\tests\GenerativeInterfacesTests.fs:line 73
Failed   FSharp.TypeProviders.SDK.Tests.GenerativeAbstractClassesTests.Abstract classes are generated correctly
Error Message:
 System.TypeLoadException : Abstract method with non-zero RVA.
Stack Trace:
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at System.Reflection.Assembly.get_ExportedTypes()
   at FSharp.TypeProviders.SDK.Tests.GenerativeAbstractClassesTests.testProvidedAssembly(FSharpFunc`2 test) in C:\Code\FSharp.TypeProviders.SDK\tests\GenerativeAbstractClassesTests.fs:line 84
```

File: tests/GenerativeAbstractClassesTests.fs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            if isVirtual then
+                //let m = ProvidedMethod(name, ps, retType)
+                let m = ProvidedMethod(name, ps, retType, invokeCode = fun args ->
+                    <@ raise (NotImplementedException(name + " is not implemented")) @>.Raw
+                    )
+                m.SetMethodAttrs (MethodAttributes.Public ||| MethodAttributes.HideBySig ||| MethodAttributes.NewSlot ||| MethodAttributes.Abstract ||| MethodAttributes.Virtual)
+                m
+            else
+                let m = ProvidedMethod(name, ps, retType)
+                m.SetMethodAttrs (MethodAttributes.Public ||| MethodAttributes.HideBySig ||| MethodAttributes.NewSlot ||| MethodAttributes.Abstract ||| MethodAttributes.Virtual)
```
Comment (sergey-tihon) [inline tests/GenerativeAbstractClassesTests.fs]: @panesofglass you add `MethodAttributes.Virtual` attribute for `not isVirtual` case, this is probably the source for 
```
 System.TypeLoadException : Abstract method with non-zero RVA.
```

**Comment by panesofglass** [comment]
Does [this](https://stackoverflow.com/questions/16877986/system-typeloadexception-was-unhandled-could-not-load-type-because-the-method-h) mean I _do_ need to provide an implementation?

File: src/ProvidedTypes.fs [hunk 3]
```diff
@@ -14477,7 +14477,7 @@ namespace ProviderImplementation.ProvidedTypes
                         match pminfo.GetInvokeCode with
                         | Some _ when ptdT.IsInterface ->
                             failwith "The provided type definition is an interface; therefore, it should not define an implementation for its members."
-                        | Some _ when pminfo.IsAbstract ->
+                        | Some _ when pminfo.IsAbstract && not pminfo.IsVirtual ->
```
Comment (panesofglass) [inline src/ProvidedTypes.fs]: I'm not certain this is the right change. I think the right combination may require some combination of provided type definition properties and `pminfo` properties.

**Comment by sergey-tihon** [comment]
We were managed to drive code to the state when all test fails with single error
```
Error Message:
   System.TypeLoadException : Abstract method with non-zero RVA.
Stack Trace:
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at System.Reflection.Assembly.get_ExportedTypes()
   at at FSharp.TypeProviders.SDK.Tests.GenerativeInterfacesTests.testProvidedAssembly(FSharpFunc`2 test) in /home/travis/build/fsprojects/FSharp.TypeProviders.SDK/tests/GenerativeInterfacesTests.fs:line 70
```

and we not sure what's exactly going wrong here. It look like code emitted incorrectly...
also there are some parts of SDK wrapped in `#if EMIT_FIELD_DATA` block that does not run now...

@dsyme we appreciate your guidance here, thx

File: tests/GenerativeAbstractClassesTests.fs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                    ProvidedParameter(name, ty))
+            if isVirtual then
+                let m = ProvidedMethod(name, ps, retType, invokeCode = fun args ->
+                    <@ raise (NotImplementedException(name + " is not implemented")) @>.Raw
+                    )
+                m.AddMethodAttrs (MethodAttributes.Virtual ||| MethodAttributes.HasSecurity)
+                m
+            else
+                let m = ProvidedMethod(name, ps, retType)
+                //m.AddMethodAttrs (MethodAttributes.Virtual ||| MethodAttributes.Abstract)
```
Comment (panesofglass) [inline tests/GenerativeAbstractClassesTests.fs]: Uncommenting this line results in both tests in this file failing in the same way as the interface generation.

File: src/ProvidedTypes.fs [hunk 5]
```diff
@@ -14740,10 +14743,12 @@ namespace ProviderImplementation.ProvidedTypes
                     match mT.GetInvokeCode with
                     | Some _ when methodBaseT.DeclaringType.IsInterface ->
                         failwith "The provided type definition is an interface; therefore, it should not define an implementation for its members."
+                    (* NOTE: These checks appear to fail for generative abstract and virtual methods.
```
Comment (panesofglass) [inline src/ProvidedTypes.fs]: These checks we added last year do not appear to work correctly for generated, virtual members.

File: src/ProvidedTypes.fsi [hunk 6]
```diff
@@ -330,6 +330,9 @@ namespace ProviderImplementation.ProvidedTypes
         /// The documentation is re-computed  every time it is required.
         member AddXmlDocComputed: xmlDocFunction: (unit -> string) -> unit
 
+        /// Add the attributes on the provided type.
+        member AddAttributes: TypeAttributes -> unit
```
Comment (panesofglass) [inline src/ProvidedTypes.fsi]: Added this member to `ProvidedTypeDefinition` in order to more easily add attributes for abstract classes. May want to move this to another ctor parameter.

**Comment by sergey-tihon** [comment]
Merged in #331 and fixed there

**Pull request closed without merging**